src/CitySearch.cpp

```diff
@@ -0,0 +1,174 @@
+ *
+ * @param filename The name of the file to read city data from.
+ * @throws std::runtime_error if the file fails to open, encounters incorrect information, or errors occur during conversion.
+ */
+void CitySearch::readCityDataFromFile(const std::string& filename) {
+	std::ifstream file(filename);
+	if (!file.is_open()) {
+		throw std::runtime_error("Failed to open the file.");
+
+	}
+
+	std::string line;
+	std::string city_name;
+
+
+	while (std::getline(file, line)) {
+		if (city_name.empty()) {
+			city_name = line;
+		}
+		else {
+			std::string coordinate_x;
+			std::string coordinate_y;
+			std::stringstream ss(line);
+			std::getline(ss, coordinate_x, '-');
+			std::getline(ss, coordinate_y, '\n');
+
+			double check_x, check_y;
+			try {
+				check_x = std::stod(coordinate_x);
+				check_y = std::stod(coordinate_y);
+			}
+			catch (const std::exception& e) {
+				throw std::runtime_error("Error converting coordinates to double: " + std::string(e.what()));
+			}
+
+			if (ss.fail() || !ss.eof()) {
+				throw std::runtime_error("Incorrect information encountered while reading the file.");
+			}
+
+
+			cities.insert({ 0, {city_name, {check_x, check_y} } });
+			city_name.clear();
+		}
+	}
+
+	file.close();
+}
+
+/**
+ * Retrieves a constant reference to the multimap of cities stored in the CitySearch object.
+ *
+ * @return A constant reference to the multimap of cities.
+ */
+const std::multimap<double, City>& CitySearch::getCities() const{
+
+	return cities;
+}
+
+/**
+ * Performs a search for cities within a specified radius from a given city, based on the selected distance norm.
+ * The search uses the coordinates of the cities and calculates the distance using the specified norm function.
+ * The resulting cities are sorted in ascending order by their distance from the given city.
+ *
+ * @param city The reference city to search around.
+ * @param radius The radius within which to search for cities.
+ * @param norm The norm to use for distance calculation (0 - L2, 1 - Linf, 2 - L1).
+ * @return A multimap of nearby cities sorted by their distance from the given city.
+ */
+std::multimap<double, City> CitySearch::search(const City& city, float radius, int norm) {
+	std::multimap<double, City> nearby_cities;
+	std::multimap<double, City> nearby_cities_by_distance;
+	double x_coordinate = city._coordinates._x_coordinate;
+	double y_coordinate = city._coordinates._y_coordinate;
+
+	// Define distance calculation function based on the selected norm
+	using DistanceFunc = std::function<double(double, double, double, double)>;
+	DistanceFunc distanceFuncs[3] = {
+		[this](double x0, double y0, double x1, double y1) { return calculateL2Distance(x0, y0, x1, y1); },
+		[this](double x0, double y0, double x1, double y1) { return calculateLInfinityDistance(x0, y0, x1, y1); },
+		[this](double x0, double y0, double x1, double y1) { return calculateL1Distance(x0, y0, x1, y1); }
+	};
+
+	DistanceFunc distanceFunc = distanceFuncs[norm];
+
```

Comment:
you are missing 1 step in finding the cities in radius, you should find vertical and horizontal rectangles, and then to find the intersection square, and to find a circle inside the square.